components/sup/src/manager.rs

```diff
@@ -416,6 +415,71 @@ pub(crate) mod sync {
         /// endpoint
         health_check_data: HashMap<ServiceGroup, HealthCheckResult>,
     }
+
+    type ManagerServicesInner = HashMap<PackageIdent, Service>;
+
+    pub struct ManagerServicesReadGuard<'a>(ReadGuard<'a, ManagerServicesInner>);
+
+    impl<'a> ManagerServicesReadGuard<'a> {
+        fn new(lock: &'a Lock<ManagerServicesInner>) -> Self { Self(lock.read()) }
+
+        pub fn services(&self) -> impl Iterator<Item = &Service> { self.0.values() }
+    }
+
+    pub struct ManagerServicesWriteGuard<'a>(WriteGuard<'a, ManagerServicesInner>);
+
+    impl<'a> ManagerServicesWriteGuard<'a> {
+        fn new(lock: &'a Lock<ManagerServicesInner>) -> Self { Self(lock.write()) }
+
+        pub fn iter(&self) -> impl Iterator<Item = (&PackageIdent, &Service)> { self.0.iter() }
+
+        pub fn insert(&mut self, key: PackageIdent, value: Service) { self.0.insert(key, value); }
+
+        pub fn remove(&mut self, key: &PackageIdent) -> Option<Service> { self.0.remove(key) }
+
+        pub fn services(&mut self) -> impl Iterator<Item = &mut Service> { self.0.values_mut() }
+
+        pub fn drain_services(&mut self) -> DrainServices<'_> {
+            DrainServices { base: self.0.drain(), }
+        }
+    }
+
+    pub struct DrainServices<'a> {
+        base: std::collections::hash_map::Drain<'a, PackageIdent, Service>,
+    }
+
+    impl<'a> Iterator for DrainServices<'a> {
+        type Item = Service;
+
+        fn next(&mut self) -> Option<Service> { self.base.next().map(|(_ident, service)| service) }
+    }
```

Comment:
Just for my own edification, why is the above pattern with `DrainServices` necessary for the `.drain()` call but not necessary for the `.services()` call?  In the latter case, simply returning `impl Iterator<Item = &mut Service>` is sufficient.  Is it because calling `.drain()` on a `HashMap` returns both key and value in the iterator, and you wanted to only return the value in this case?  Or is there something else here that I'm not seeing?

Reply:
I couldn't figure out a way to make it work with returning a trait because it needs to include a lifetime parameter that isn't part of the `Iterator` trait. I wanted to keep the details abstract, so rather than expose the implementation by returning `std::collections::hash_map::Drain`, the simplest thing seemed to be wrapping it in a newtype. Since that gave the opportunity to return only the `Service` to the iterator and simplify the consumer code, I figured "why not?"

Reply:
I _think_ that makes sense 😁

Reply:
The first implementation I tried was:
```rust
pub fn drain(&mut self) -> impl Iterator<Item = (PackageIdent, Service)> { self.0.drain() }
```
which gives the following errors:
```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
   --> components/sup/src/manager.rs:452:84
    |
452 |         pub fn drain(&mut self) -> impl Iterator<Item = (PackageIdent, Service)> { self.0.drain() }
    |                                                                                    ^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the impl at 441:10...
   --> components/sup/src/manager.rs:441:10
    |
441 |     impl<'a> ManagerServicesWriteGuard<'a> {
    |          ^^
note: ...so that the reference type `&mut lock_api::rwlock::RwLockWriteGuard<'a, parking_lot::raw_rwlock::RawRwLock, std::collections::HashMap<habitat_core::package::ident::PackageIdent, manager::service::Service>>` does not outlive the data it points at
   --> components/sup/src/manager.rs:452:84
    |
452 |         pub fn drain(&mut self) -> impl Iterator<Item = (PackageIdent, Service)> { self.0.drain() }
    |                                                                                    ^^^^^^
    = note: but, the lifetime must be valid for the static lifetime...
note: ...so that return value is valid for the call
   --> components/sup/src/manager.rs:452:36
    |
452 |         pub fn drain(&mut self) -> impl Iterator<Item = (PackageIdent, Service)> { self.0.drain() }
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

You'll notice that in the stdlib, all the Iterator types are concrete, not `impl Iterator`; I think that's for lifetime issues like these.